rpc-broker/src/policy.c

```diff
@@ -0,0 +1,258 @@
+{
+    int rule_idx;
+    FILE *policy_fh;
+    size_t rbytes, line_length;
+    char *line;
+    char current_rule[RULE_MAX_LENGTH] = { 0 };
+    struct rule *current;
+
+    rule_idx = 0;
+    etc->count = 0;
+
+    policy_fh = fopen(rule_filepath, "r");
+    if (!policy_fh) {
+        DBUS_BROKER_WARNING("/etc policy stat of file <%s> failed %s",
+                             rule_filepath, strerror(errno));
+        return;
+    }
+
+    line = NULL;
+
+    while ((getline(&line, &rbytes, policy_fh) > 0) && rule_idx < MAX_RULES) {
+
+        if (rbytes > RULE_MAX_LENGTH - 1) {
+            DBUS_BROKER_WARNING("Invalid policy rule %zu exceeds max-rule",
+                                                                  rbytes);
+        } else if (line && isalpha(line[0])) {
+            line_length = strlen(line);
+            line[line_length - 1] = '\0';
+            memcpy(current_rule, line, rbytes);
+            current = &(etc->rules[rule_idx]);
+            create_rule(current, current_rule) < 0 ? free_rule(*current) :
+                                                               rule_idx++;
+        }
+
+        if (line)
+            free(line);
+
+        line = NULL;
+    }
+
+    if (line)
+        free(line);
+
+    etc->count = rule_idx;
+    fclose(policy_fh);
+}
+
+struct policy *build_policy(const char *rule_filename)
+{
+    struct policy *dbus_policy;
+    struct etc_policy *etc;
+    int dom_idx;
+    DBusMessage *vms;
+    DBusConnection *conn;
+    DBusMessageIter iter, sub;
+    void *arg;
+    struct domain_policy *current;
+    char uuid[64];
+
+    dbus_policy = calloc(1, sizeof *dbus_policy);
+    etc = &(dbus_policy->etc);
+    build_etc_policy(etc, rule_filename);
+    dbus_policy->domain_count= 0;
+
+    dom_idx = 0;
+    vms = db_list();
+    if (!vms)
+        return dbus_policy;
+    conn = create_dbus_connection();
+
+    dbus_message_iter_init(vms, &iter);
+    dbus_message_iter_recurse(&iter, &sub);
+
+
+    while (dbus_message_iter_get_arg_type(&sub) != DBUS_TYPE_INVALID) {
+
+        dbus_message_iter_get_basic(&sub, &arg);
+        current = &(dbus_policy->domains[dom_idx]);
+        strcpy(current->uuid, arg);
+        // alter the uuid from underscores to dashes
+        // dbus returns underscores while the vm db uses dashes
+        // if not the vm db policy check on strcmp on the uuid is off
+        TRANSFORM_UUID(arg, uuid);
+        strcpy(current->uuid_db_fmt, uuid);
+
+        errno = 0;
+        if (errno != 0) {
+            DBUS_BROKER_WARNING("Domain ID error <%s>", strerror(errno));
+            continue;
+        }
+
+        get_rules(conn, current);
+
+        dbus_message_iter_next(&sub);
+        dom_idx++;
+    }
+
+    dbus_policy->domain_count = dom_idx;
+    dbus_message_unref(vms);
+    dbus_connection_unref(conn);
+
+    return dbus_policy;
+}
+
+void free_rule(struct rule r)
+{
+    if (r.destination)
+        free((char *) (r.destination));
```

Comment:
No need for casts in this function.

Reply:
It doesn't compile without the cast, `free` complains.